File: Linked List/Python/LC 142 - LINKED LIST CYCLE II.py
# LEET CODE 142 - Linked List Cycle II
# QUESTION LINK - https://leetcode.com/problems/linked-list-cycle-ii/

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return None

        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                temp = head
                while temp != slow:
                    temp = temp.next
                    slow = slow.next
                return slow

        return None

File: Linked List/Java/LC 142 - LINKED LIST CYCLE II.java
// LEET CODE 142 - Linked List Cycle II
// QUESTION LINK - https://leetcode.com/problems/linked-list-cycle-ii/

class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) return null;

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if (slow == fast) {
                ListNode temp = head;
                while (temp != slow) {
                    temp = temp.next;
                    slow = slow.next;
                }
                return slow;
            }
        }

        return null;
    }
}
